Introduce the ability to add new OSDs using a device list

fmount opened the pull request:
This change just adds a new set of tasks to
the tripleo_cluster_osds role which automates
the osds zap action.
This is really useful for some environments
that come up w/ a not clean disk map, resulting
in a locked device tree which prevents osds to
be added in the ceph cluster.

Closes #12

Signed-off-by: Francesco Pantano <fpantano@redhat.com>

fmount committed:
Add zap tasks on tripleo_cluster_osds role

This change just adds a new set of tasks to
the tripleo_cluster_osds role which automates
the osds zap action.
This is really useful for some environments
that come up w/ a not clean disk map, resulting
in a locked device tree which prevents osds to
be added in the ceph cluster.

Closes #1

Signed-off-by: Francesco Pantano <fpantano@redhat.com>

fultonj commented:
This is handy for development but I don't think we should have this in production. Too dangerous as it could lead to data loss. Instead the hardware should be cleaned before deployment by ironic as described here:

  https://bugzilla.redhat.com/show_bug.cgi?id=1613918

The openstack/ceph team added a feature just like this in puppet-ceph:

https://github.com/openstack/puppet-ceph/commit/108f5500adb38f5e8c2ddf271937ef184f57067a

and then we removed it:

https://github.com/openstack/puppet-ceph/commit/9f52c306e1584d8def0890002ad12df87117e882

Becasue we were worried it would delete someone's data. Better to know you never delete data in your deployment tools.

fmount commented:
> This is handy for development but I don't think we should have this in production. Too dangerous as it could lead to data loss. Instead the hardware should be cleaned before deployment by ironic as described here:

Ack and agree: the focus is day1 and hw is supposed to be clean.
Let's keep this PR open for now and see if it's useful to turn it into something else.
> 
> https://bugzilla.redhat.com/show_bug.cgi?id=1613918
> 
> The openstack/ceph team added a feature just like this in puppet-ceph:
> 
> [openstack/puppet-ceph@108f550](https://github.com/openstack/puppet-ceph/commit/108f5500adb38f5e8c2ddf271937ef184f57067a)
> 
> and then we removed it:
> 
> [openstack/puppet-ceph@9f52c30](https://github.com/openstack/puppet-ceph/commit/9f52c306e1584d8def0890002ad12df87117e882)
> 
> Becasue we were worried it would delete someone's data. Better to know you never delete data in your deployment tools.

sebastian-philipp commented on roles/tripleo_cluster_osds/tasks/zap.yml:
```diff
@@ -0,0 +1,15 @@
+  when: "{{ device_list | length == 0 }}"
+
+- name: Run 'ceph orch zap'
+  shell: "{{ ceph_cli }} orch device zap {{ item[0] }} {{ item[1] }} --force"
```
note, this is relates to https://tracker.ceph.com/issues/45792

fmount replied:
> note, this is relates to https://tracker.ceph.com/issues/45792

Hi,
thanks for pointing out to the tracker! 
This definitely explains the issue on my env!
I like the `--all-available-devices` feature because on day2  operators shouldn't care about running commands/playbooks/scripts to add their new devices on the cluster.
Regarding the device zapping, since the scope of the project is  `focus on day1`, describe your cluster and push the button, this role will be removed or turned into something else (as per https://github.com/fmount/tripleo-ceph/issues/12#issuecomment-637086465 in the OpenStack context ironic cleaning should be sufficient) but I'm still using it for dev purposed until [1] is solved.

[1] https://tracker.ceph.com/issues/45792

sebastian-philipp replied:
Thanks for the explanation!

fmount committed:
Introduce the ability to add new OSD(s) using a device list

This commit just adds a set of tasks to build a list of
devices and add a new OSD via orchestrator using the
<host>:<devices|join(',') syntax.

Signed-off-by: Francesco Pantano <fpantano@redhat.com>
